### PR TITLE
Add HTMLTextDocument class

### PR DESCRIPTION
**Pull request by wch**
The `HTMLTextDocument` class has a similar purpose to `HTMLDocument`, except that its contents are loaded from an HTML file.

It is meant to be used with https://github.com/posit-dev/py-shiny/pull/746

This PR also makes it possible change how HTML dependencies are rendered, by setting the following:

```py
htmltools.html_dependency_render_mode = "json"
```

Normally HTML dependencies are not visible when a `Tag`/`TagList` calls `.render()`. However, if that value is set to `"json"`, then HTML dependencies will be rendered into the document as something like this (note that the JSON has been prettified here, but normally is not):

```
<script type="application/json" data-html-dependency="">
{
  "name":"strftime",
  "version":"0.9.2",
  "source":{
    "package":"shiny",
    "subdir":"www/shared/strftime/"
  },
  "script":[
    {
      "src":"strftime-min.js"
    }
  ],
  "stylesheet":[    
  ],
  "meta":[    
  ],
  "all_files":false,
  "head":null
}
</script>
```

Then when the HTML is read in, the `HTMLDependency` object can be re-created like this:

```py
args = json.loads(x)
dep = HTMLDependency(**args)
```

### CODE REVIEW

File: htmltools/_core.py [hunk 1]
```diff
@@ -358,17 +361,28 @@ def show(self, renderer: Literal["auto", "ipython", "browser"] = "auto") -> obje
...
 
-    def __str__(self) -> str:
-        return str(self.get_html_string())
-
     def __eq__(self, other: Any) -> bool:
         return _equals_impl(self, other)
 
+    def __str__(self) -> str:
+        rendered = self.render()
+        res = rendered["html"]
```
Comment (jcheng5) [inline htmltools/_core.py:369]: This should be a function

Comment (jcheng5) [reply htmltools/_core.py:369]: Maybe add a comment that this is for Quarto integration

**Review by jcheng5 — COMMENTED** [review]
Other than refactoring that duplicated code into a function, LGTM

**Review by jcheng5 — APPROVED** [review]

File: htmltools/_core.py (hunk 1 above)
Comment (wch) [reply htmltools/_core.py]: Done in abab13d.

**Pull request merged**